File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/multirt/MultiRTConfigDialog.java
package net.kdt.pojavlaunch.multirt;

import android.annotation.SuppressLint;
import android.content.Context;
import android.widget.Button;

import androidx.activity.result.ActivityResultLauncher;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.arata.yukarilauncher.R;

public class MultiRTConfigDialog {
    private AlertDialog mDialog;
    private RecyclerView mDialogView;

    /**
     * Show the dialog, refreshes the adapter data before showing it
     */
    public void show() {
        refresh();
        mDialog.show();
    }

    @SuppressLint("NotifyDataSetChanged")
    //only used to completely refresh the list, it is necessary
    public void refresh() {
        RecyclerView.Adapter<?> adapter = mDialogView.getAdapter();
        if (adapter != null) adapter.notifyDataSetChanged();
    }

    /**
     * Build the dialog behavior and style
     */
    public void prepare(Context activity, ActivityResultLauncher<Object> installJvmLauncher) {
        mDialogView = new RecyclerView(activity);
        mDialogView.setLayoutManager(new LinearLayoutManager(activity, LinearLayoutManager.VERTICAL, false));
        RTRecyclerViewAdapter adapter = new RTRecyclerViewAdapter(MultiRTUtils.getRuntimes());
        mDialogView.setAdapter(adapter);

        mDialog = new AlertDialog.Builder(activity, R.style.CustomAlertDialogTheme)
                .setTitle(R.string.multirt_config_title)
                .setView(mDialogView)
                .setPositiveButton(R.string.multirt_import, (dialog, which) -> installJvmLauncher.launch(null))
                .setNeutralButton(R.string.multirt_delete_runtime, null)
                .create();

        // Custom button behavior without dismiss
        mDialog.setOnShowListener(dialog -> {
            Button button = ((AlertDialog) dialog).getButton(AlertDialog.BUTTON_NEUTRAL);
            button.setOnClickListener(view -> {
                boolean isEditing = !adapter.getIsEditing();
                adapter.setIsEditing(isEditing);
                button.setText(isEditing ? R.string.multirt_swap_setdefault : R.string.multirt_delete_runtime);
            });
        });
    }
}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/utils/runtime/MultiRTConfigDialog.kt
package com.arata.yukarilauncher.utils.runtime

import android.annotation.SuppressLint
import android.content.Context
import android.widget.Button
import androidx.activity.result.ActivityResultLauncher
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.arata.yukarilauncher.R

/** マルチランタイム設定ダイアログを管理するクラス。 */
class MultiRTConfigDialog {
    /** 設定ダイアログのインスタンス。 */
    private var mDialog: AlertDialog? = null
    /** ダイアログ内のRecyclerView。 */
    private var mDialogView: RecyclerView? = null

    /** ダイアログを表示する。 */
    fun show() {
        refresh()
        mDialog?.show()
    }

    /** アダプターのデータを更新して再描画する。 */
    @SuppressLint("NotifyDataSetChanged")
    fun refresh() {
        val adapter = mDialogView?.adapter
        adapter?.notifyDataSetChanged()
    }

    /**
     * ダイアログを構築する。
     * @param activity コンテキスト
     * @param installJvmLauncher JVMインストール用 ActivityResultLauncher
     */
    fun prepare(activity: Context, installJvmLauncher: ActivityResultLauncher<Any?>) {
        mDialogView = RecyclerView(activity).apply {
            layoutManager = LinearLayoutManager(activity, LinearLayoutManager.VERTICAL, false)
            adapter = RTRecyclerViewAdapter(MultiRTUtils.runtimes)
        }

        mDialog = AlertDialog.Builder(activity, R.style.CustomAlertDialogTheme)
            .setTitle(R.string.multirt_config_title)
            .setView(mDialogView)
            .setPositiveButton(R.string.multirt_import) { _, _ -> installJvmLauncher.launch(Any()) }
            .setNeutralButton(R.string.multirt_delete_runtime, null)
            .create()

        mDialog?.setOnShowListener { dialog ->
            val button = (dialog as AlertDialog).getButton(AlertDialog.BUTTON_NEUTRAL)
            button.setOnClickListener {
                val adapter = mDialogView?.adapter as RTRecyclerViewAdapter
                val isEditing = !adapter.isEditing
                adapter.isEditing = isEditing
                button.setText(if (isEditing) R.string.multirt_swap_setdefault else R.string.multirt_delete_runtime)
            }
        }
    }
}